File: java/253-meeting-rooms-ii.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
class Solution {
    public int minMeetingRooms(Interval[] intervals) {
        if (intervals == null || intervals.length == 0) {
            return 0;
        }
        Arrays.sort(intervals, (i1, i2) -> {
            return Integer.compare(i1.start, i2.start);
        });
        PriorityQueue<Interval> heap = new PriorityQueue<>((i1, i2) -> {
            return Integer.compare(i1.end, i2.end);
        });
        for (Interval interval : intervals) {
            if (heap.isEmpty()) {
                heap.offer(interval);
            } else {
                Interval curr = heap.peek();
                if (curr.end > interval.start) {
                    heap.offer(interval);
                } else {
                    curr.end = interval.end;
                    heap.poll();
                    heap.offer(curr);
                }
            }
        }
        return heap.size();
    }
}

File: python3/253.meeting-rooms-ii.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        if intervals is None or len(intervals) == 0:
            return 0
        points = []
        for interval in intervals:
            points.append(Point(interval.start, True))
            points.append(Point(interval.end, False))
        points = sorted(points, key=lambda p: (p.pos, p.isStart))
        ans = 0
        room = 0
        for point in points:
            if point.isStart:
                room += 1
            else:
                room -= 1
            ans = max(ans, room)
        return ans
    
class Point:
    def __init__(self, pos, isStart):
        self.pos = pos
        self.isStart = isStart
    def __repr__(self):
        return "(%s, %s)" % (self.pos, self.isStart)